A71-xds-fallback.md

```diff
@@ -0,0 +1,242 @@
+xDS server is unavailable. xDS support for client-side gRPC is described
+in gRFC [A27][A27] and server-side support is described in [A36][A36].
+
+## Background
+
+xDS (also known as xDS Discovery Service) is a suite of APIs for discovering
+and subscribing to the configuration of a server mesh. Even a brief downtime
+of the xDS control plane may cause significant disruption in the service mesh
+inter-component connectivity and cause wider outages.
+
+The current xDS implementation in gRPC uses a single global instance of
+the XdsClient class to communicate with the xDS control plane. This instance
+is shared across all channels and xDS-enabled gRPC server listeners to reduce
+overhead by providing a shared cache for xDS resources and reducing the number
+of connections to xDS servers.
+
+### Reservations about using the fallback server data
+
+We are expecting to use this in an environment where the data provided by
+the secondary xDS server is better than nothing but is less desirable than
+the data in the primary server. E.g. a fallback environment that uses
+an outdated static configuration. This means that while we definitely want
+to use the fallback data in preference to the client simply failing, we will
+always prefer to use the data from the primary server over the data from
+the fallback server. In principle, this means that whenever we have a resource
+already cached from the primary server, we want to use that instead of falling
+back.
+
+We have no guarantee that a combination of resources from different xDS servers
+form a valid cohesive configuration, so we cannot make this determination on
+a per-resource basis. We need any given gRPC channel or server listener to only
+use the resources from a single server.
+
+### Related Proposals: 
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A36: xDS-Enabled Servers][A36]
+* [A40: xDS Configuration Dump via Client Status Discovery Service in gRPC][A40]
+* [A47: xDS Federation][A47]
+* [A57: XdsClient Failure Mode Behavior][A57]
+
+## Proposal
+
+Implement the following changes
+
+1. Change global XdsClient from a single instance to a map of instances keyed
+by channel target.
+1. Update CSDS to aggregate configs from multiple XdsClient instances.
+1. Change xDS bootstrap code to return the full list of servers for each
+xDS authority from the bootstrap config.
+1. Change XdsClient to support fallback within each xDS authority.
+
+### Change global XdsClient from a single instance to per-target instances
+
+Current gRPC implementation is using a single XdsClient for all data plane
+targets. For fallback, if any one channel triggers fallback, then all channels
+regardless of authority would switch to using the data from the fallback xDS
+control plane, which is not desirable.
+
+gRPC servers using the xDS configuration will share the same XdsClient instance
+keyed with a dedicated well-known key value.
+
+Changes XdsClient to per-data plane target would enable gRPC to switch to
```

Comment:
This looks to be missing a subject. Was this intended to be "Changing XdsClient..."

That's unclear though. Is that a theoretical, "if we made this change then we'd get this result," or is it "we _are_ making this change to get this result." Maybe something like:

> gRPC Channel will share XdsClient instances keyed by the data plane target , which would enable..."

And then maybe the "This change would change..." paragraph gets merged in here. Right now it is unclear what "this change" is referring to.

Reply:
Please take a loog at #407 where I tried addressing your comment.